src/jit/codegenarmarch.cpp

```diff
@@ -880,6 +886,216 @@ void CodeGen::genPutArgReg(GenTreeOp* tree)
+
+        CORINFO_CLASS_HANDLE objClass = source->gtObj.gtClass;
+
+        structSize = compiler->info.compCompHnd->getClassSize(objClass);
+        isHfa      = compiler->IsHfa(objClass);
+        gcPtrs     = new (compiler, CMK_Codegen) BYTE[treeNode->gtNumRegs + treeNode->gtNumSlots];
+        gcPtrCount = compiler->info.compCompHnd->getClassGClayout(objClass, &gcPtrs[0]);
+    }
+
+    // If we have an HFA we can't have any GC pointers,
+    // if not then the max size for the the struct is 16 bytes
+    if (isHfa)
+    {
+        noway_assert(gcPtrCount == 0);
+    }
+
+    unsigned varNumOut    = compiler->lvaOutgoingArgSpaceVar;
+    unsigned argOffsetMax = compiler->lvaOutgoingArgSpaceSize;
+    unsigned argOffsetOut = treeNode->gtSlotNum * TARGET_POINTER_SIZE;
+
+    // Put on stack first
+    unsigned nextIndex     = treeNode->gtNumRegs;
+    unsigned structOffset  = nextIndex * TARGET_POINTER_SIZE;
+    int      remainingSize = structSize - structOffset;
+
+    while (remainingSize >= TARGET_POINTER_SIZE)
+    {
+        var_types type = compiler->getJitGCType(gcPtrs[nextIndex]);
+
+        if (varNode != nullptr)
+        {
+            // Load from our varNumImp source
+            emit->emitIns_R_S(INS_ldr, emitTypeSize(type), baseReg, varNumInp, structOffset);
+        }
+        else
+        {
+            // check for case of destroying the addrRegister while we still need it
+            assert(baseReg != addrReg);
+
+            // Load from our address expression source
+            emit->emitIns_R_R_I(INS_ldr, emitTypeSize(type), baseReg, addrReg, structOffset);
+        }
+
+        // Emit str instruction to store the register into the outgoing argument area
+        emit->emitIns_S_R(INS_str, emitTypeSize(type), baseReg, varNumOut, argOffsetOut);
+        argOffsetOut += TARGET_POINTER_SIZE;  // We stored 4-bytes of the struct
+        assert(argOffsetOut <= argOffsetMax); // We can't write beyound the outgoing area area
+        remainingSize -= TARGET_POINTER_SIZE; // We loaded 4-bytes of the struct
+        structOffset += TARGET_POINTER_SIZE;
+        nextIndex += 1;
+    }
+
+    while (remainingSize > 0)
+    {
+        {
+            int loadSize  = remainingSize;
+            remainingSize = 0;
+
+            // We should never have to do a non-pointer sized load when we have a LclVar source
+            assert(varNode == nullptr);
+
+            // the left over size is smaller than a pointer and thus can never be a GC type
+            assert(varTypeIsGC(compiler->getJitGCType(gcPtrs[nextIndex])) == false);
+
+            var_types loadType = TYP_UINT;
+            if (loadSize == 1)
+            {
+                loadType = TYP_UBYTE;
+            }
+            else if (loadSize == 2)
+            {
+                loadType = TYP_USHORT;
+            }
+            else
+            {
+                // Need to handle additional loadSize cases here
+                noway_assert(loadSize == 4);
```

Comment:
I think this assertion is needless for ARM32. This assertion comes form the implementation of `CodeGen::genPutArgStk. I'll remove this and fix the assertion in `CodeGen::genPutArgStk` later.